// @mathigon/core/build/core.js
'use strict';

Object.defineProperty(exports, '__esModule', { value: true });

// =============================================================================
// Core.js | Utilities
// (c) Mathigon
// =============================================================================



/**
 * Empty Function.
 */
function noop() {}


/**
 * Creates a random UID string of a given length.
 * @param {?number} n
 * @returns {string}
 */
function uid(n = 10) {
  return Math.random().toString(36).substr(2, n);
}


/**
 * If obj is a function, it evaluates it with a given set of attributes.
 * Otherwise it just returns obj directly.
 * @param {Function|*} obj
 * @param {...*} args
 * @returns {*}
 */
function run(obj, ...args) {
  if (obj instanceof Function) return obj(...args);
  return obj;
}


/**
 * Checks if x is strictly equal to any one of the following arguments
 * @param {*} x
 * @param {...*} values
 * @returns {boolean}
 */
function isOneOf(x, ...values) {
  for (let v of values) {
    if (x === v) return true;
  }
  return false;
}


/**
 * Merges multiple objects into the first one.
 * @param {Object} first
 * @param {...Object} others
 */
function extend(first, ...others) {
  for (let obj of others) {
    for (let key in obj) {
      if (obj.hasOwnProperty(key)) first[key] = obj[key];
    }
  }
}


/**
 * Deep extends obj1 using th
 * @param obj1
 * @param obj2
 */
function deepExtend(obj1, obj2) {
  for (let i of Object.keys(obj2)) {
    if (i in obj1 && Array.isArray(obj1[i]) && Array.isArray(obj2[i])) {
      obj1[i] = obj1[i].concat(obj2[i]);
    } else if (i in obj1 && obj1[i] instanceof Object &&
                            obj2[i] instanceof Object) {
      deepExtend(obj1[i], obj2[i]);
    } else {
      obj1[i] = obj2[i];
    }
  }
}


/**
 * Bounds a number between a lower and an upper limit.
 * @param {number} x
 * @param {number} min
 * @param {number} max
 * @returns {number}
 */
function clamp(x, min = -Infinity, max = Infinity) {
  return Math.min(max, Math.max(min, x));
}


/**
 * Checks the average speed of a function by running it n times.
 * @param {Function} fn
 * @param {?number} n
 * @returns {number}
 */
function performance(fn, n = 100) {
  window.performance.clearMarks();
  window.performance.clearMeasures();

  window.performance.mark('start');
  for (let i=0; i<n; ++i) fn();
  window.performance.mark('end');

  window.performance.measure('time', 'start', 'end');
  let t = window.performance.getEntriesByName('time')[0].duration;
  return t/n;
}


/**
 * Squares a number.
 * @param {number} x
 * @returns {number}
 */
function square(x) {
  return x * x;
}


/**
 * Cubes a number.
 * @param {number} x
 * @returns {number}
 */
function cube(x) {
  return x * x * x;
}


/**
 * Replacement for setTimeout() that is synchronous for time 0.
 * @param {Function} fn
 * @param {?number} t
 * @returns {?number}
 */
function delay(fn, t = 0) {
  if (t) { return setTimeout(fn, t); } else { fn(); }
}


/**
 * Returns a promise that resolves after a fixed time.
 * @param {number} t
 * @returns {Promise}
 */
function wait(t) {
  return new Promise(resolve => setTimeout(resolve, t));
}


/**
 * Creates a new promise together with functions to resolve or reject.
 * @returns {{promise: Promise, resolve: Function, reject: Function}}
 */
function defer() {
  let resolve, reject;

  const promise = new Promise((_resolve, _reject) => {
    resolve = _resolve;
    reject = _reject;
  });

  // This prevents exceptions when promises without .catch are rejected:
  promise.catch(function(error) { return error; });

  return { promise, resolve, reject };
}


/**
 * Checks if an object contains a property.
 * @param {Object} obj
 * @param {number} key
 * @returns {boolean}
 */
function has(obj, key) {
  return Object.prototype.hasOwnProperty.call(obj, key);
}


/**
 * @param {Object} collection
 * @param {Function} fn
 * @returns {Object}
 */
function each(collection, fn) {
  let newCollection = Array.isArray(collection) ? [] : {};
  let keys = Object.keys(collection);
  for (let k of keys) {
    newCollection[k] = fn(collection[k], k);
  }
  return newCollection;
}


/**
 * @param {Object} collection
 * @param {Function} fn
 * @returns {*}
 */
function some(collection, fn) {
  if (Array.isArray(collection)) {
    for (let i = 0; i < collection.length; ++i) {
      let x = fn(collection[i], i);
      if (x != null) return x;
    }
  } else {
    for (let k of Object.keys(collection)) {
      let x = fn(collection[k], k);
      if (x != null) return x;
    }
  }
}


/**
 * Function wrapper that modifies a function to cache its return values. This
 * is useful for performance intensive functions which are called repeatedly
 * with the same arguments. However it can reduce performance for functions
 * which are always called with different arguments. Note that argument
 * comparison doesn't not work with Objects or nested arrays.

 * @param {Function} fn
 * @returns {Function}
 */
function cache(fn) {
  let cached = new Map();
  return function(...args) {
    let argString = args.join('--');
    if (!cached.has(argString)) cached.set(argString, fn(...args));
    return cached.get(argString);
  };
}


/**
 * Function wrapper that prevents a function from being executed more than once
 * every t ms. This is particularly useful for optimising callbacks for
 * continues events like scroll, resize or slider move.
 *
 * @param {Function} fn
 * @param {?number} t
 * @returns {Function}
 */
function throttle(fn, t = 0) {
  let delay = false;
  let repeat = false;
  return function(...args) {
    if (delay) {
      repeat = true;
    } else {
      fn.apply(null, args);
      delay = true;
      setTimeout(function() {
        if (repeat) fn(...args);
        delay = false;
        repeat = false;
      }, t);
    }
  };
}


/**
 * Shallow copies any JavaScript object.
 * @param {*} obj
 * @returns {*}
 */
function shallowCopy(obj) {
  // Handle (simple) strings, numbers, booleans, null and undefined
  let type = typeof obj;
  if (obj == null || isOneOf(type, 'number', 'string', 'boolean')) return obj;

  // Handle other type objects
  if (obj instanceof Number)  return +obj;
  if (obj instanceof String)  return '' + obj;
  if (obj instanceof Boolean) return !!obj;
  if (obj instanceof Date)    return new Date(obj.valueOf());
  if (obj instanceof RegExp)  return new RegExp(obj);

  // Handle Arrays and Objects
  return each(obj, x => x);
}


let DEEP_COPY_STORE = null;

function deepCopyHelper(obj) {
  // Handle (simple) strings, numbers, booleans, null and undefined
  let type = typeof obj;
  if (obj == null || isOneOf(type, 'number', 'string', 'boolean')) return obj;

  // Handle other type objects
  if (obj instanceof Number)  return +obj;
  if (obj instanceof String)  return '' + obj;
  if (obj instanceof Boolean) return !!obj;
  if (obj instanceof Date)    return new Date(obj.valueOf());
  if (obj instanceof RegExp)  return new RegExp(obj);

  // Fallback
  if (!(obj instanceof Object)) return obj;

  // Avoids Recursive Loops
  if (DEEP_COPY_STORE.has(obj)) return DEEP_COPY_STORE.get(obj);

  // Handle Arrays and Objects
  let copy = Array.isArray(obj) ? [] : {};
  DEEP_COPY_STORE.set(obj, copy);

  let keys = Object.keys(obj);
  for (let k of keys) copy[k] = deepCopyHelper(obj[k]);

  return copy;
}

/**
 * Deep copies any JavaScript object.
 * @param {*} obj
 * @returns {*}
 */
function deepCopy(obj) {
  DEEP_COPY_STORE = new WeakMap();
  let copy = deepCopyHelper(obj);
  DEEP_COPY_STORE = null;
  return copy;
}

// =============================================================================


function _tabulateWith(fn, vals, args) {
  if (!args.length) return run(fn, ...vals);

  let newArgs = args.slice(0);
  let x = newArgs.shift();

  let result = [];
  for (let i=0; i<x; ++i) {
    result.push(_tabulateWith(fn, vals.concat([i]), newArgs));
  }
  return result;
}

/**
 * Creates a multi-dimensional array contains fn(i1, i2, ..., in) in cell
 * [i1][i2]...[in]. If the first argument is not a function, every item in the
 * resulting array will have that static value.
 *
 * @param {Function|*} fn
 * @param {...number} dimensions
 * @returns {Array}
 */
function tabulate(fn, ...dimensions) {
  return _tabulateWith(fn, [], dimensions);
}


/**
 * @param {number} a
 * @param {?number} b
 * @param {?number} step
 * @returns {Array}
 */
function list(a, b, step = 1) {
  let arr = [];

  if (b == null && a >= 0) {
    for (let i=0; i<a; i += step) arr.push(i);
  } else if (b == null) {
    for (let i=0; i>a; i -= step) arr.push(i);
  } else if (a <= b) {
    for (let i=a; i<=b; i += step) arr.push(i);
  } else {
    for (let i=a; i>=b; i -= step) arr.push(i);
  }

  return arr;
}


/**
 * Returns the last item in an array, or the ith item from the end.
 * @param {Array} array
 * @param {?Number} i
 * @returns {*}
 */
function last(array, i=0) {
  return array[array.length - 1 - i];
}


/**
 * Maps a function over multiple arrays, the entries of which are passed as
 * multiple arguments.
 * @param {Function} fn
 * @param {...Array} args
 * @returns {Array}
 */
function map(fn, ...args) {
  let length = Math.max(...args.map(a => a.length));
  return tabulate(i => fn(...args.map(a => a[i])), length);
}


/**
 * Finds the sum of all elements in an numeric array. This operation is safe,
 * i.e. any values that can't be cast to a number are counted as 0.
 * @param {Array} array
 * @returns {number}
 */
function total(array) {
  if (!array.length) return 0;
  return +array.reduce((t, v) => t + (+v || 0));
}


/**
 * Evaluates the average of all the elements in an array.
 * @param {Array} array
 * @returns {number}
 */
function average(array) {
  return array.total / array.length;
}


/**
 * Checks if an array contains a specific value.
 * @param {Array} array
 * @param {*} value
 * @returns {boolean}
 */
function contains(array, value) {
  return array.indexOf(value) >= 0;
}


/**
 *
 * @param array
 * @param id
 */
function extract(array, id) {
  return array.map(a => a[id]);
}


/**
 *
 * @param keys
 * @param values
 * @returns {{}}
 */
function zip(keys, values) {
  let obj = {};
  for (let i=0; i < keys.length; ++i)
    obj[keys[i]] = values[i];
  return obj;
}


/**
 *
 * @param {Array} array
 * @param {string} id
 * @param {?boolean} reverse
 * @returns {Array}
 */
function sortBy(array, id, reverse = false) {
  return array.slice(0).sort((a, b) => {
    return a[id] < b[id] ? (reverse ? 1 : -1) :
      a[id] > b[id] ? (reverse ? -1 : 1) : 0;
  });
}


/**
 *
 * @param {Array} array
 * @param {Function} fn
 * @param {?boolean} reverse
 * @returns {Array}
 */
function sortByFn(array, fn, reverse = false) {
  return array.slice(0).sort((a, b) => {
    let x = fn(a);
    let y = fn(b);
    return x < y ? (reverse ? 1 : -1) : x > y ? (reverse ? -1 : 1) : 0;
  });
}


/**
 * Returns a function that can be called repeatedly, and returns items of the
 * array, continuously looping
 * @param {Array} array
 * @returns {Function}
 */
function loop(array) {
  let i = 0;
  return () => array[(i++) % array.length];
}


/**
 *
 * @param array
 * @returns {Function}
 */
function unique(array) {
  return array.filter((a, i, _this) => _this.indexOf(a) == i);
  // return [...new Set([this])];
}


/**
 * Removes any null or undefined values in array a.
 * @param {Array} array
 * @return {Array}
 */
function clean(array) {
  return array.filter(a => a != null);
}


/**
 * Removes all occurrences of x from an array.
 * @param {Array} array
 * @param {*} x
 * @return {Array}
 */
function without(array, x) {
  return array.filter(a => a !== x);
}


/**
 *
 * @param array
 */
function flatten(array) {
  return array.reduce((a, b) =>
    a.concat(Array.isArray(b) ? flatten(b) : b), []);
}


/**
 *
 * @param array
 */
function cumulative(array) {
  let total = 0;
  return array.map(a => total += a);
}


/**
 * Breaks an array into chunks of size at most n.
 * @param array
 * @param n
 * @returns {Array}
 */
function chunk(array, n) {
  let chunks = [];
  for (let i = 0; i < array.length; i += n) {
    chunks.push(array.slice(i, i + n));
  }
  return chunks;
}


/**
 * Rotates the elements of an array by offset.
 * @param array
 * @param offset
 */
function rotate(array, offset) {
  let n = array.length;
  offset = ((offset % n) + n) % n; // offset could initially be negative...

  let start = array.slice(0, offset);
  let end = array.slice(offset);
  return end.concat(start);
}


/**
 * Returns all elements that are in both a1 and a2.
 * @param a1
 * @param a2
 * @returns {Array}
 */
function intersect(a1, a2) {
  let result = [];
  for (let i = 0; i < a1.length; ++i) {
    if (a2.indexOf(a1[i]) === -1) result.push(a1[i]);
  }
  return result;
}


/**
 * Returns all elements that are only in one of a1 and a2.
 * @param a1
 * @param a2
 * @returns {Array}
 */
function difference(a1, a2) {
  let notIn1 = a2.filter(a => !a1.includes(a));
  let notIn2 = a1.filter(a => !a2.includes(a));
  return [...notIn1, ...notIn2];
}


/**
 * Converts an array into a one-way linked list, with .val and .next properties.
 * @param array
 * @returns {Array}
 */
function toLinkedList(array) {
  const result = array.map(a => ({val: a, next: null}));
  const n = result.length;

  for (let i = 0; i < n; ++i) {
    result[i].next = result[(i + 1) % n];
  }

  return result;
}

// =============================================================================
// Core.js | Evented
// (c) Mathigon
// =============================================================================



function process(events, options) {
  if (options.lowercase) events = events.toLowerCase();
  return events.split(options.split).map(e => e.trim());
}

/**
 * Base class for event management. It is rarely used on its own, but many other
 * classes inherit from Evented.
 */
class Evented {

  constructor({ split = ' ', lowercase = false } = {}) {
    this._options = { split, lowercase };
    this._events = {};
  }

  /**
   * Adds an event listener for one or more events.
   * @param {string} events
   * @param {Function} fn
   */
  on(events, fn) {
    for (let e of process(events, this._options)) {
      if (!(e in this._events)) this._events[e] = [];
      this._events[e].push(fn);
    }
  }

  /**
   * Adds a one-time event listener to one or more events.
   * @param {string} events
   * @param {Function} fn
   */
  one(events, fn) {
    let _this = this;
    function callback(...args) {
      _this.off(events, callback);
      fn(...args);
    }
    this.on(events, callback);

  }

  /**
   * Removes an event listener from one or more events.
   * @param {string} events
   * @param {Function} fn
   */
  off(events, fn) {
    for (let e of process(events, this._options)) {
      if (e in this._events)
        this._events[e] = this._events[e].filter(x => x !== fn);
    }
  }

  /**
   * Triggers one or more events, and executes all bound event listeners with
   * the given arguments.
   * @param {string} events
   * @param {...*} args
   */
  trigger(events, ...args) {
    for (let e of process(events, this._options)) {
      if (e in this._events)
        this._events[e].forEach(function(fn) { fn.apply(this, args); });
    }
  }

}

// =============================================================================


/**
 * Splits a string into space separated words.
 * @param {string} str
 * @returns {Array<string>}
 */
function words(str) {
  if (!str) return [];
  return str.trim().split(/\s+/);
}


/**
 * Converts a string to title case.
 * @param {string} str
 * @returns {string}
 */
function toTitleCase(str) {
  return str.replace(/\S+/g, a =>  a.charAt(0).toUpperCase() + a.slice(1));
}


/**
 * Converts a string to camel case.
 * @param {string} str
 * @returns {string}
 */
function toCamelCase(str) {
  return str.toLowerCase().replace(/^-/,'')
                          .replace(/-(.)/g, (match, g) => g.toUpperCase());
}


/**
 * Repeats a string n times.
 * @param {string} str
 * @param {?number} n
 * @returns {string}
 */
function repeat(str, n = 1) {
  for (let i=1; i<n; ++i) str += str;
  return str;
}


/**
 * Checks if a string is a palindrome.
 * @param {string} str
 * @returns {boolean}
 */
function isPalindrome(str) {
  return str === str.split('').reverse().join('');
}


/**
 * Determines the Levenshtein distance between two strings.
 * @param {string} s1
 * @param {string} s2
 * @returns {number}
 */
function stringDistance(s1, s2) {
  let arr = tabulate(0, s1.length + 1, s2.length + 1);
  for (let i = 0; i <= s1.length; i++) arr[i][0] = i;
  for (let i = 0; i <= s2.length; i++) arr[0][i] = i;

  for (let i = 1; i <= s1.length; i++)
    for (let j = 1; j <= s2.length; j++)
      arr[i][j] = Math.min(arr[i - 1][j - 1] +
                           (s1.charAt(i - 1) === s2.charAt(j - 1) ? 0 : 1),
                           arr[i - 1][j] + 1, arr[i][j - 1] + 1);

  return arr[s1.length][s2.length];
}


/**
 * Tries to autocorrect a word from a dictionary.
 * @param {string} word
 * @param {Array<string>} dict
 * @returns {string|null}
 */
function autocorrect(word, dict) {
  let maxDistance = word.length / 2;
  let distances = dict.map(w => ({w, d: stringDistance(word, w)}))
                      .filter(({d}) => d < maxDistance);
  let bestMatch = sortBy(distances, 'd')[0];
  return bestMatch ? bestMatch.w : null;
}

// =============================================================================
// Core.js | Types
// (c) Mathigon
// =============================================================================



const typeRegexp = /\s([a-zA-Z]+)/;

/**
 * Returns the type of an object. Possible values are 'number', 'string',
 * 'boolean', 'array', 'object', 'date', 'function', 'NaN', 'null', 'undefined',
 * as well as class names like 'htmldocument'.
 * @param {*} obj
 * @returns {string}
 */
function typeOf(obj) {
  if (obj == null) return '' + obj;
  if (Number.isNaN(obj)) return 'NaN';

  return toString.call(obj).match(typeRegexp)[1].toLowerCase();
}

/**
 * Checks if an object has a specific type.
 * @param {*} x
 * @param {string} type
 * @returns {boolean}
 */
function isType(x, type) {
  return typeOf(x) === type;
}

/*
 * Checks if an object is a string.
 * @param {*} x
 * @returns {boolean}
 */
function isString(x) {
  return (x instanceof String) || (typeof x === 'string');
}

/*
 * Checks if an object is a number.
 * @param {*} x
 * @returns {boolean}
 */
function isNumber(x) {
  return (x instanceof Number) || (typeof x === 'number');
}

/*
 * Checks if an object is an integer.
 * @param {*} x
 * @returns {boolean}
 */
function isInteger(x) {
  return x % 1 === 0;
}

/*
 * Checks if an object is an array.
 * @param {*} x
 * @returns {boolean}
 */
function isArray(x) {
  return Array.isArray(x);
}

/*
 * Checks if an object is a date.
 * @param {*} x
 * @returns {boolean}
 */
function isDate(x) {
  return Object.prototype.toString.call(x) === '[object Date]';
}

/*
 * Checks if an object is a boolean.
 * @param {*} x
 * @returns {boolean}
 */
function isBoolean(x) {
  return (x instanceof Boolean) || (typeof x === 'boolean');
}

/*
 * Checks if an object is an object.
 * @param {*} x
 * @returns {boolean}
 */
function isObject(x) {
  return x === Object(x);
}

// =============================================================================

exports.tabulate = tabulate;
exports.list = list;
exports.last = last;
exports.map = map;
exports.total = total;
exports.average = average;
exports.contains = contains;
exports.extract = extract;
exports.zip = zip;
exports.sortBy = sortBy;
exports.sortByFn = sortByFn;
exports.loop = loop;
exports.unique = unique;
exports.clean = clean;
exports.without = without;
exports.flatten = flatten;
exports.cumulative = cumulative;
exports.chunk = chunk;
exports.rotate = rotate;
exports.intersect = intersect;
exports.difference = difference;
exports.toLinkedList = toLinkedList;
exports.Evented = Evented;
exports.words = words;
exports.toTitleCase = toTitleCase;
exports.toCamelCase = toCamelCase;
exports.repeat = repeat;
exports.isPalindrome = isPalindrome;
exports.stringDistance = stringDistance;
exports.autocorrect = autocorrect;
exports.typeOf = typeOf;
exports.isType = isType;
exports.isString = isString;
exports.isNumber = isNumber;
exports.isInteger = isInteger;
exports.isArray = isArray;
exports.isDate = isDate;
exports.isBoolean = isBoolean;
exports.isObject = isObject;
exports.noop = noop;
exports.uid = uid;
exports.run = run;
exports.isOneOf = isOneOf;
exports.extend = extend;
exports.deepExtend = deepExtend;
exports.clamp = clamp;
exports.performance = performance;
exports.square = square;
exports.cube = cube;
exports.delay = delay;
exports.wait = wait;
exports.defer = defer;
exports.has = has;
exports.each = each;
exports.some = some;
exports.cache = cache;
exports.throttle = throttle;
exports.shallowCopy = shallowCopy;
exports.deepCopy = deepCopy;
